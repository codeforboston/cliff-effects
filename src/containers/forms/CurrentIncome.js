import { connect } from 'react-redux';

import { CurrentIncomeStep } from '../../forms/CurrentIncome';

import { setCashValue } from '../../actions';

const mapStateToProps = (state) => {
  return {
    currentClient: state.getIn([
      'client',
      'current', 
    ]),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setIncomeValue({ name, value }) {
      dispatch(
        setCashValue({
          time: 'current',
          name,
          value,
        })
      );
    },
  };
};

const CurrentIncomeStepContainer = connect(mapStateToProps, mapDispatchToProps)(CurrentIncomeStep);

export { CurrentIncomeStepContainer as CurrentIncomeStep };
