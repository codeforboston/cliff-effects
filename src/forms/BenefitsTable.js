// REACT COMPONENTS
import React from 'react';
import {
  Table
} from 'semantic-ui-react';

// CUSTOM COMPONENTS
// Both the table and graph should just be added to a results page, but
// this will do for now

// BENEFIT LOGIC
import { getSNAPBenefits } from '../programs/federal/snap';
import { getHousingBenefit } from '../programs/state/massachusetts/housing';

// OBJECT MANIPULATION
import { cloneDeep } from 'lodash';


const getSignSymbol = function ( num ) {
  if ( num > 0 ) { return '+'; }
  else if ( num < 0 ) { return '-'; }
  else { return ''; }
};  // End getSignSymbol()


const BenefitsTable = function ( props ) {

  var currentClient = cloneDeep( props.client ),
      futureClient  = cloneDeep( currentClient );
  // Hack because everything's using future value instead
  // of just using current value.
  currentClient.future.earned = currentClient.current.earned;




  var SNAPBenefitCurrent  = Math.round( getSNAPBenefits( currentClient ).benefitValue * 12 ),
      SNAPBenefitFuture   = Math.round( getSNAPBenefits( futureClient ).benefitValue * 12 ),
      SNAPDiff            = SNAPBenefitFuture - SNAPBenefitCurrent,
      sec8BenefitCurrent  = Math.round( getHousingBenefit( currentClient ).benefitValue * 12 ),
      sec8BenefitFuture   = Math.round( getHousingBenefit( futureClient ).benefitValue * 12 ),
      sec8Diff            = sec8BenefitFuture - sec8BenefitCurrent,
      totalBenefitCurrent = SNAPBenefitCurrent + sec8BenefitCurrent,
      totalBenefitFuture  = SNAPBenefitFuture + sec8BenefitFuture,
      totalDiff           = SNAPDiff + sec8Diff,
      incomeCurrent       = Math.round( currentClient.current.earned * 12 ),
      incomeFuture        = Math.round( futureClient.future.earned * 12 ),
      incomeDiff          = incomeFuture - incomeCurrent,
      netCurrent          = totalBenefitCurrent + incomeCurrent,
      netFuture           = totalBenefitFuture + incomeFuture,
      netDiff             = totalDiff + incomeDiff;


const   columnHeaderStyle = {
                            background: 'rgba(0, 181, 173, 1)',
                            color: 'white',
                            fontSize: '1.3em',
                            fontWeight: 900,
                            textAlign: 'center',
                            borderRadius: 'inherit',
                            letterSpacing: '0.02em',
                            }
      , totalsRowStyle    = {
                            borderTop: "2px solid rgba(0, 181, 173, 1)",
                            fontWeight: 700,
                            fontSize: '1.1em',
                            padingTop: '0.25em'
                              }
      , rowHeaderStyle    = {
                    fontSize: '1.1em',
                    fontWeight: 700,
                    textAlign: 'left'
                    }
      , totalsRowHeaderStyle = {
                    fontSize: '1.2em',
                    fontWeight: 700,
                    textAlign: 'left',
                    borderTop: "2px solid rgba(0, 181, 173, 1)",
                    padingTop: '0.25em'


      };






const SNAPBenefitRow = function( props ){
  if(!currentClient.current.hasSnap ) return (null)
    return (
      <Table.Row>
        <Table.Cell style={rowHeaderStyle}>SNAP</Table.Cell>
        <Table.Cell textAlign='right'>${SNAPBenefitCurrent} / year</Table.Cell>
        <Table.Cell textAlign='right'>${SNAPBenefitFuture} / year</Table.Cell>
        <Table.Cell textAlign='right'>{ getSignSymbol(SNAPDiff) } ${Math.abs(SNAPDiff)} / year</Table.Cell>
      </Table.Row>
    )
  };

const Sec8BenefitRow  = function( props ){  
  if(!currentClient.current.hasHousing) return (null)
    return (
      <Table.Row>
        <Table.Cell style={rowHeaderStyle}>Section 8 Housing</Table.Cell>
        <Table.Cell textAlign='right'>${sec8BenefitCurrent} / year</Table.Cell>
        <Table.Cell textAlign='right'>${sec8BenefitFuture} / year</Table.Cell>
        <Table.Cell textAlign='right'>{ getSignSymbol(sec8Diff) } ${Math.abs(sec8Diff)} / year</Table.Cell>
      </Table.Row>
    ) 
};

const TotalBenefitsRow = function( props ){
  if(!currentClient.current.hasSnap || !currentClient.current.hasHousing) return (null)
    return(
      <Table.Row>
        <Table.Cell textAlign='right' width={3} style={totalsRowHeaderStyle}>Total Benefits</Table.Cell>
        <Table.Cell textAlign='right' width={3} style={totalsRowStyle}>${totalBenefitCurrent} / year</Table.Cell>
        <Table.Cell textAlign='right' width={3} style={totalsRowStyle}>${totalBenefitFuture} / year</Table.Cell>
        <Table.Cell textAlign='right' width={3} style={totalsRowStyle}>{ getSignSymbol(totalDiff) } ${Math.abs(totalDiff)} / year</Table.Cell>
      </Table.Row>
    )
};

const IncomeRow = function ( props ) { 
    return (
      <Table.Row>
        <Table.Cell style={rowHeaderStyle}>Income</Table.Cell>
        <Table.Cell textAlign='right'>${incomeCurrent} / year</Table.Cell>
        <Table.Cell textAlign='right'>${incomeFuture} / year</Table.Cell>
        <Table.Cell textAlign='right'>{ getSignSymbol(incomeDiff) } ${Math.abs(incomeDiff)} / year</Table.Cell>
      </Table.Row>
    )
};

const TotalsRow = function ( props ) {
    return (
      <Table.Row style={{border: 'none'}}>
        <Table.Cell textAlign='right' width={3} style={totalsRowHeaderStyle}>Net Total</Table.Cell>
        <Table.Cell textAlign='right' width={3} style={totalsRowStyle}>${netCurrent} / year</Table.Cell>
        <Table.Cell textAlign='right' width={3} style={totalsRowStyle}>${netFuture} / year</Table.Cell>
        <Table.Cell textAlign='right' width={3} style={totalsRowStyle}>{ getSignSymbol(netDiff) } ${Math.abs(netDiff)} / year</Table.Cell>
      </Table.Row>
    )
  };


  return (
    <wrapper>
      <Table celled>
       <Table.Header>
          <Table.Row >
            <Table.Cell style={columnHeaderStyle} width={3}>Benefit</Table.Cell>
            <Table.Cell style={columnHeaderStyle} width={3}>Current Benefits</Table.Cell>
            <Table.Cell style={columnHeaderStyle} width={3}>New Estimate</Table.Cell>
            <Table.Cell style={columnHeaderStyle} width={3}>Difference</Table.Cell>
          </Table.Row>
      </Table.Header>
        <Table.Body>
          <SNAPBenefitRow client={currentClient} />
          <Sec8BenefitRow client={currentClient} />
          <TotalBenefitsRow client={currentClient} />
          <IncomeRow />
          <TotalsRow />
        </Table.Body>
      </Table>
    </wrapper>
  )

};  // End BenefitsTable(<>)


export {
	BenefitsTable
};
