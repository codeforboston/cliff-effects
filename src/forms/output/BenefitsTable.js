// REACT COMPONENTS
import React from 'react';
import { Table } from 'semantic-ui-react';

// BENEFIT LOGIC
import { applyAndPushBenefits } from '../../programs/applyAndPushBenefits';

// OBJECT MANIPULATION
import { cloneDeep } from 'lodash';


const getSignSymbol = function (num) {
  if (num > 0) {
    return '+';
  }
  else if (num < 0) {
    return '-';
  }
  else { return ''; }
};  // End getSignSymbol()


const BenefitsTable = function ({ client, snippets }) {

  var clone = cloneDeep(client);
  var curr = clone.current;

  var allData         = {},
      activeBenefits  = [ `income` ];

  if (curr.hasSection8) {
    activeBenefits.push(`section8`);
  }

  if (curr.hasSnap) {
    activeBenefits.push(`snap`);
  }

  var currentCalcData = {
    activeBenefits: activeBenefits,
    dataToAddTo:    allData,
    clientToChange: clone,
    timeframe:      `current`,
  };
  applyAndPushBenefits (currentCalcData);

  // Add to the `current` data already there
  var futureCalcData = {
    activeBenefits: activeBenefits,
    dataToAddTo:    allData,
    clientToChange: clone,
    timeframe:      `future`,
  };
  applyAndPushBenefits (futureCalcData);

  // @todo Abstract getting values for each row
  var income   = allData.income,
      section8 = allData.section8,
      snap     = allData.snap;

  var sec8BenefitCurrent = 0,
      sec8BenefitFuture  = 0,
      SNAPBenefitCurrent = 0,
      SNAPBenefitFuture  = 0;

  if (curr.hasSection8) {
    sec8BenefitCurrent     = Math.round(section8[ 0 ]);
    sec8BenefitFuture      = Math.round(section8[ 1 ]);
  }

  if (curr.hasSnap) {
    SNAPBenefitCurrent = Math.round(snap[ 0 ]);
    SNAPBenefitFuture  = Math.round(snap[ 1 ]);
  }

  var SNAPDiff            = SNAPBenefitFuture - SNAPBenefitCurrent,
      sec8Diff            = sec8BenefitFuture - sec8BenefitCurrent,
      totalBenefitCurrent = SNAPBenefitCurrent + sec8BenefitCurrent,
      totalBenefitFuture  = SNAPBenefitFuture + sec8BenefitFuture,
      totalDiff           = SNAPDiff + sec8Diff,
      incomeCurrent       = Math.round(income[ 0 ]),
      incomeFuture        = Math.round(income[ 1 ]),
      incomeDiff          = incomeFuture - incomeCurrent,
      netCurrent          = totalBenefitCurrent + incomeCurrent,
      netFuture           = totalBenefitFuture + incomeFuture,
      netDiff             = totalDiff + incomeDiff;

  /** @todo: linting - discuss indentation for object properties and colons */
  const columnHeaderStyle = {
          background:    'rgba(0, 181, 173, 1)',
          color:         'white',
          fontSize:      '1.3em',
          fontWeight:    900,
          textAlign:     'center',
          borderRadius:  'inherit',
          letterSpacing: '0.02em',
        }
        , totalsRowStyle    = {
          borderTop:  '2px solid rgba(0, 181, 173, 1)',
          fontWeight: 700,
          fontSize:   '1.1em',
          padingTop:  '0.25em',
        }
        , rowHeaderStyle    = {
          fontSize:   '1.1em',
          fontWeight: 700,
          textAlign:  'left',
        }
        , totalsRowHeaderStyle = {
          fontSize:   '1.2em',
          fontWeight: 700,
          textAlign:  'left',
          borderTop:  '2px solid rgba(0, 181, 173, 1)',
          padingTop:  '0.25em',


        };


  const SNAPBenefitRow = function({ client, snippets }){

    if (!client.current.hasSnap) {
      return (null);
    }

    return (
      <Table.Row>
        <Table.Cell style={ rowHeaderStyle }>{ snippets.i_rowSNAP }</Table.Cell>
        <Table.Cell textAlign='right'>{ snippets.i_dollarSign }{SNAPBenefitCurrent}{ snippets.i_perMonth }</Table.Cell>
        <Table.Cell textAlign='right'>{ snippets.i_dollarSign }{SNAPBenefitFuture}{ snippets.i_perMonth }</Table.Cell>
        <Table.Cell textAlign='right'>{ getSignSymbol(SNAPDiff) } { snippets.i_dollarSign }{ Math.abs(SNAPDiff) }{ snippets.i_perMonth }</Table.Cell>
      </Table.Row>
    );
  };

  const Sec8BenefitRow  = function({ client, snippets }){
    if (!client.current.hasSection8) {
      return (null);
    }

    return (
      <Table.Row>
        <Table.Cell style={ rowHeaderStyle }>{ snippets.i_rowSection8 }</Table.Cell>
        <Table.Cell textAlign='right'>{ snippets.i_dollarSign }{sec8BenefitCurrent}{ snippets.i_perMonth }</Table.Cell>
        <Table.Cell textAlign='right'>{ snippets.i_dollarSign }{sec8BenefitFuture}{ snippets.i_perMonth }</Table.Cell>
        <Table.Cell textAlign='right'>{ getSignSymbol(sec8Diff) } { snippets.i_dollarSign }{ Math.abs(sec8Diff) }{ snippets.i_perMonth }</Table.Cell>
      </Table.Row>
    );
  };

  const TotalBenefitsRow = function({ client, snippets }){
    if (!client.current.hasSnap || !client.current.hasSection8) {
      return (null);
    }

    return (
      <Table.Row>
        <Table.Cell
          textAlign='right'
          width={ 3 }
          style={ totalsRowHeaderStyle }>{ snippets.i_rowTotalBenefits }
        </Table.Cell>
        <Table.Cell
          textAlign='right'
          width={ 3 }
          style={ totalsRowStyle }>{ snippets.i_dollarSign }{totalBenefitCurrent}{ snippets.i_perMonth }
        </Table.Cell>
        <Table.Cell
          textAlign='right'
          width={ 3 }
          style={ totalsRowStyle }>{ snippets.i_dollarSign }{totalBenefitFuture}{ snippets.i_perMonth }
        </Table.Cell>
        <Table.Cell
          textAlign='right'
          width={ 3 }
          style={ totalsRowStyle }>{ getSignSymbol(totalDiff) } { snippets.i_dollarSign }{ Math.abs(totalDiff) }{ snippets.i_perMonth }
        </Table.Cell>
      </Table.Row>
    );
  };

  const IncomeRow = function ({ snippets }) {
    return (
      <Table.Row>
        <Table.Cell style={ rowHeaderStyle }>{ snippets.i_rowIncome }</Table.Cell>
        <Table.Cell textAlign='right'>{ snippets.i_dollarSign }{incomeCurrent}{ snippets.i_perMonth }</Table.Cell>
        <Table.Cell textAlign='right'>{ snippets.i_dollarSign }{incomeFuture}{ snippets.i_perMonth }</Table.Cell>
        <Table.Cell textAlign='right'>{ getSignSymbol(incomeDiff) } { snippets.i_dollarSign }{ Math.abs(incomeDiff) }{ snippets.i_perMonth }</Table.Cell>
      </Table.Row>
    );
  };

  const TotalsRow = function ({ snippets }) {
    return (
      <Table.Row style={{ border: 'none' }}>
        <Table.Cell
          textAlign='right'
          width={ 3 }
          style={ totalsRowHeaderStyle }>{ snippets.i_rowNetTotal }
        </Table.Cell>
        <Table.Cell
          textAlign='right'
          width={ 3 }
          style={ totalsRowStyle }>{ snippets.i_dollarSign }{netCurrent}{ snippets.i_perMonth }
        </Table.Cell>
        <Table.Cell
          textAlign='right'
          width={ 3 }
          style={ totalsRowStyle }>{ snippets.i_dollarSign }{netFuture}{ snippets.i_perMonth }
        </Table.Cell>
        <Table.Cell
          textAlign='right'
          width={ 3 }
          style={ totalsRowStyle }>{ getSignSymbol(netDiff) } { snippets.i_dollarSign }{ Math.abs(netDiff) }{ snippets.i_perMonth }
        </Table.Cell>
      </Table.Row>
    );
  };

  return (
    <div>
      <Table celled>
        <Table.Header>
          <Table.Row >
            <Table.Cell
              style={ columnHeaderStyle }
              width={ 3 }>{ snippets.i_columnBenefit }
            </Table.Cell>
            <Table.Cell
              style={ columnHeaderStyle }
              width={ 3 }>{ snippets.i_columnCurrentBenefits }
            </Table.Cell>
            <Table.Cell
              style={ columnHeaderStyle }
              width={ 3 }>{ snippets.i_columnNewEstimate }
            </Table.Cell>
            <Table.Cell
              style={ columnHeaderStyle }
              width={ 3 }>{ snippets.i_columnDifference }
            </Table.Cell>
          </Table.Row>
        </Table.Header>
        <Table.Body>
          <SNAPBenefitRow 
            client={ clone }
            snippets={ snippets } />
          <Sec8BenefitRow 
            client={ clone }
            snippets={ snippets } />
          <TotalBenefitsRow 
            client={ clone } 
            snippets={ snippets } />
          <IncomeRow snippets={ snippets } />
          <TotalsRow snippets={ snippets } />
        </Table.Body>
      </Table>
    </div>
  );

};  // End BenefitsTable(<>)


export { BenefitsTable };
