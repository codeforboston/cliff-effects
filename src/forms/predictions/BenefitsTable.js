// REACT COMPONENTS
import React from 'react';
import { Table } from 'semantic-ui-react';

// BENEFIT LOGIC
import { applyAndPushBenefits } from '../../benefits/applyAndPushBenefits';

// OBJECT MANIPULATION
import { cloneDeep } from 'lodash';


// @todo Move to utils file somewhere?
const getSignSymbol = function (num) {
  if (num > 0) {
    return `+`;
  } else if (num < 0) {
    return `-`;
  } else { return ``; }
};


const BenefitsTable = function ({ client, translations }) {
  const clone = cloneDeep(client);
  const curr = clone.current;

  let allData         = {},
      activeBenefits  = [
        `earned`,
        ...curr.benefits,
      ];

  let currentCalcData = {
    activeBenefits: activeBenefits,
    dataToAddTo:    allData,
    clientToChange: clone,
    timeframe:      `current`,
    USState:        client.USState,
  };
  applyAndPushBenefits (currentCalcData);

  // Add to the `current` data already there
  let futureCalcData = {
    activeBenefits: activeBenefits,
    dataToAddTo:    allData,
    clientToChange: clone,
    timeframe:      `future`,
    USState:        client.USState,
  };
  applyAndPushBenefits (futureCalcData);

  let earned = allData.earned;
  
  let currentBenefits     = {},
      futureBenefits      = {},
      benefitDiffs        = {},
      totalDiff           = 0,
      totalBenefitCurrent = 0,
      totalBenefitFuture  = 0;

  for (let benefitIndex = 0; benefitIndex < curr.benefits.length; benefitIndex++) {
    
    let benefit       = curr.benefits[ benefitIndex ],
        benefitData = allData[ benefit ];
    
    if (benefitData) {
      let [
        currentBenefit,
        futureBenefit, 
      ] = benefitData;
      
      currentBenefits[ benefit ] = Math.round(currentBenefit);
      futureBenefits[ benefit ]  = Math.round(futureBenefit);

      totalBenefitCurrent += currentBenefits[ benefit ];
      totalBenefitFuture  += futureBenefits[ benefit ];
    } else {
      currentBenefits[ benefit ] = 0;
      futureBenefits[ benefit ]  = 0;
    }

    benefitDiffs[ benefit ] = futureBenefits[ benefit ] - currentBenefits[ benefit ];
    totalDiff += benefitDiffs[ benefit ];
  }

  let earnedCurrent = Math.round(earned[ 0 ]),
      earnedFuture  = Math.round(earned[ 1 ]),
      earnedDiff    = earnedFuture - earnedCurrent,
      netCurrent    = totalBenefitCurrent + earnedCurrent,
      netFuture     = totalBenefitFuture + earnedFuture,
      netDiff       = totalDiff + earnedDiff;

  // @todo Possible to break the following components out?
  let columnHeaderStyle = {
        background:    `rgba(0, 181, 173, 1)`,
        color:         `white`,
        fontSize:      `1.3em`,
        fontWeight:    900,
        textAlign:     `center`,
        borderRadius:  `inherit`,
        letterSpacing: `0.02em`,
      },
      totalsRowStyle = {
        borderTop:  `2px solid rgba(0, 181, 173, 1)`,
        fontWeight: 700,
        fontSize:   `1.1em`,
        padingTop:  `0.25em`,
      },
      rowHeaderStyle = {
        fontSize:   `1.1em`,
        fontWeight: 700,
        textAlign:  `left`,
      },
      totalsRowHeaderStyle = {
        fontSize:   `1.2em`,
        fontWeight: 700,
        textAlign:  `left`,
        borderTop:  `2px solid rgba(0, 181, 173, 1)`,
        padingTop:  `0.25em`,
      };

  const TotalBenefitsRow = function({ client, translations }){
    if (client.current.benefits.length <= 1) {
      return (null);
    }

    return (
      <Table.Row>
        <Table.Cell
          textAlign = { `right` }
          width     = { 3 }
          style     = { totalsRowHeaderStyle }>
          { translations.i_rowTotalBenefits }
        </Table.Cell>
        <Table.Cell
          textAlign = { `right` }
          width     = { 3 }
          style     = { totalsRowStyle }>
          { translations.i_beforeMoneyWithTime }{totalBenefitCurrent}{ translations.i_afterMoneyWithTime }
        </Table.Cell>
        <Table.Cell
          textAlign = { `right` }
          width     = { 3 }
          style     = { totalsRowStyle }>
          { translations.i_beforeMoneyWithTime }{totalBenefitFuture}{ translations.i_afterMoneyWithTime }
        </Table.Cell>
        <Table.Cell
          textAlign = { `right` }
          width     = { 3 }
          style     = { totalsRowStyle }>
          { getSignSymbol(totalDiff) } { translations.i_beforeMoneyWithTime }{ Math.abs(totalDiff) }{ translations.i_afterMoneyWithTime }
        </Table.Cell>
      </Table.Row>
    );
  };

  const EarnedRow = function ({ translations }) {
    return (
      <Table.Row>
        <Table.Cell style={ rowHeaderStyle }>{ translations.i_rowEarned }</Table.Cell>
        <Table.Cell textAlign={ `right` }>{ translations.i_beforeMoneyWithTime }{earnedCurrent}{ translations.i_afterMoneyWithTime }</Table.Cell>
        <Table.Cell textAlign={ `right` }>{ translations.i_beforeMoneyWithTime }{earnedFuture}{ translations.i_afterMoneyWithTime }</Table.Cell>
        <Table.Cell textAlign={ `right` }>{ getSignSymbol(earnedDiff) } { translations.i_beforeMoneyWithTime }{ Math.abs(earnedDiff) }{ translations.i_afterMoneyWithTime }</Table.Cell>
      </Table.Row>
    );
  };

  const TotalsRow = function ({ translations }) {
    return (
      <Table.Row style={{ border: `none` }}>
        <Table.Cell
          textAlign = { `right` }
          width     = { 3 }
          style     = { totalsRowHeaderStyle }>
          { translations.i_rowNetTotal }
        </Table.Cell>
        <Table.Cell
          textAlign = { `right` }
          width     = { 3 }
          style     = { totalsRowStyle }>
          { translations.i_beforeMoneyWithTime }{netCurrent}{ translations.i_afterMoneyWithTime }
        </Table.Cell>
        <Table.Cell
          textAlign = { `right` }
          width     = { 3 }
          style     = { totalsRowStyle }>
          { translations.i_beforeMoneyWithTime }{netFuture}{ translations.i_afterMoneyWithTime }
        </Table.Cell>
        <Table.Cell
          textAlign = { `right` }
          width     = { 3 }
          style     = { totalsRowStyle }>
          { getSignSymbol(netDiff) } { translations.i_beforeMoneyWithTime }{ Math.abs(netDiff) }{ translations.i_afterMoneyWithTime }
        </Table.Cell>
      </Table.Row>
    );
  };

  const benefitRows = [];

  for (let benefitIndex = 0; benefitIndex < curr.benefits.length; benefitIndex++) {
    let benefit = curr.benefits[ benefitIndex ],
        diff    = benefitDiffs[ benefit ],
        label   = translations[ `i_row_${benefit}` ];

    benefitRows.push(
      <Table.Row
        key={ benefit }>
        <Table.Cell style={ rowHeaderStyle }>{ label }</Table.Cell>
        <Table.Cell textAlign={ `right` }>{ translations.i_beforeMoneyWithTime }{currentBenefits[ benefit ]}{ translations.i_afterMoneyWithTime }</Table.Cell>
        <Table.Cell textAlign={ `right` }>{ translations.i_beforeMoneyWithTime }{futureBenefits[ benefit ]}{ translations.i_afterMoneyWithTime }</Table.Cell>
        <Table.Cell textAlign={ `right` }>{ getSignSymbol(diff) } { translations.i_beforeMoneyWithTime }{ Math.abs(diff) }{ translations.i_afterMoneyWithTime }</Table.Cell>
      </Table.Row>
    );
  }

  return (
    <Table celled>
      <Table.Header>
        <Table.Row >
          <Table.Cell
            style = { columnHeaderStyle }
            width = { 3 }>
            { translations.i_columnBenefit }
          </Table.Cell>
          <Table.Cell
            style = { columnHeaderStyle }
            width = { 3 }>
            { translations.i_columnCurrentBenefits }
          </Table.Cell>
          <Table.Cell
            style = { columnHeaderStyle }
            width = { 3 }>
            { translations.i_columnNewEstimate }
          </Table.Cell>
          <Table.Cell
            style = { columnHeaderStyle }
            width = { 3 }>
            { translations.i_columnDifference }
          </Table.Cell>
        </Table.Row>
      </Table.Header>

      <Table.Body>
        { benefitRows }
        <TotalBenefitsRow 
          client       = { clone } 
          translations = { translations } />
        <EarnedRow translations={ translations } />
        <TotalsRow translations={ translations } />
      </Table.Body>
    </Table>
  );

};  // Ends <BenefitsTable>


export { BenefitsTable };
