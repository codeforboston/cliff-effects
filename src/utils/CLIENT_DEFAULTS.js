import { cloneDeep } from 'lodash';

/**
 * Default client values.
 * MUST ALWAYS BE VALID
 * 
 * Note: A 'positive number' money value can have more than
 *     two decimal places.
 */
const CLIENT_DEFAULTS = {

  // Property Name                | Default Value   | Valid Value Description

  current: {
    // Current programs
    hasSnap:                        false,            // boolean
    hasHousing:                     false,            // boolean
    // Household
    // 'm_' for 'member'
    household:  [{                  // 'head' member  // one or more member objects
        m_age:                      30,               // whole number
        m_role:                     'head',           // 'head', 'spouse', 'member'
        m_disabled:                 false             // boolean
    }],
    // Income
    earned:                         0,                // positive number (can have more than two decimal places)
    TAFDC:                          0,                // positive number
    SSI:                            0,                // positive number
    SSDI:                           0,                // positive number
    childSupportIn:                 0,                // positive number
    unemployment:                   0,                // positive number
    workersComp:                    0,                // positive number
    pension:                        0,                // positive number
    socialSecurity:                 0,                // positive number
    alimony:                        0,                // positive number
    otherIncome:                    0,                // positive number
    incomeExclusions:               0,                // positive number
    // Expenses
    childDirectCare:                0,                // positive number
    childBeforeAndAfterSchoolCare:  0,                // positive number
    childTransportation:            0,                // positive number
    childOtherCare:                 0,                // positive number
    earnedBecauseOfChildCare:       0,                // positive number
    childSupportPaidOut:            0,                // positive number
    adultDirectCare:                0,                // positive number
    adultTransportation:            0,                // positive number
    adultOtherCare:                 0,                // positive number
    disabledAssistance:             0,                // positive number
    earnedBecauseOfAdultCare:       0,                // positive number
    disabledMedical:                0,                // positive number
    otherMedical:                   0,                // positive number
    /** @todo When client has section 8, switch this to 'housingVoucher' */
    /** @todo Make shelter expense values more robust. */
    shelter:                        'homeless',       // 'homeless', 'homeowner', 'renter', 'housingVoucher'
    contractRent:                   0,                // positive number
    rentShare:                      0,                // positive number
    rent:                           0,                // positive number
    mortgage:                       0,                // positive number
    housingInsurance:               0,                // positive number
    propertyTax:                    0,                // positive number
    climateControl:                 false,            // boolean
    nonHeatElectricity:             false,            // boolean
    phone:                          false,            // boolean
    fuelAssistance:                 false,            // boolean  Note: Input is 'yes', 'no', but should be being converted
    otherExpenses:                  0                 // positive number
  },

  // Note: A `future` value will default to the client's `current` value until a user changes it explicitly
  future: {},                       // same structure // same

};

CLIENT_DEFAULTS.future = cloneDeep( CLIENT_DEFAULTS.current );


export { CLIENT_DEFAULTS };
