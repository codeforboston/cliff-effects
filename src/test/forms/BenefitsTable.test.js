import React from 'react';
import renderer from 'react-test-renderer';

import { BenefitsTable } from '../../forms/BenefitsTable';

import { CLIENT_DEFAULTS } from '../../utils/CLIENT_DEFAULTS';

test('Benefits table renders correctly', () => {
  const rendered = renderer.create(
    <BenefitsTable client={CLIENT_DEFAULTS} />
  );
  expect(rendered.toJSON()).toMatchSnapshot();
});