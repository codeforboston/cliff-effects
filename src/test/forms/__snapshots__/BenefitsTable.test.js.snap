// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Benefits table renders correctly 1`] = `
<wrapper>
  <table
    className="ui celled table"
  >
    <thead
      className=""
    >
      <tr
        className=""
      >
        <td
          className="three wide"
          style={
            Object {
              "background": "rgba(0, 181, 173, 1)",
              "borderRadius": "inherit",
              "color": "white",
              "fontSize": "1.3em",
              "fontWeight": 900,
              "letterSpacing": "0.02em",
              "textAlign": "center",
            }
          }
        >
          Benefit
        </td>
        <td
          className="three wide"
          style={
            Object {
              "background": "rgba(0, 181, 173, 1)",
              "borderRadius": "inherit",
              "color": "white",
              "fontSize": "1.3em",
              "fontWeight": 900,
              "letterSpacing": "0.02em",
              "textAlign": "center",
            }
          }
        >
          Current Benefits
        </td>
        <td
          className="three wide"
          style={
            Object {
              "background": "rgba(0, 181, 173, 1)",
              "borderRadius": "inherit",
              "color": "white",
              "fontSize": "1.3em",
              "fontWeight": 900,
              "letterSpacing": "0.02em",
              "textAlign": "center",
            }
          }
        >
          New Estimate
        </td>
        <td
          className="three wide"
          style={
            Object {
              "background": "rgba(0, 181, 173, 1)",
              "borderRadius": "inherit",
              "color": "white",
              "fontSize": "1.3em",
              "fontWeight": 900,
              "letterSpacing": "0.02em",
              "textAlign": "center",
            }
          }
        >
          Difference
        </td>
      </tr>
    </thead>
    <tbody
      className=""
    >
      <tr
        className=""
      >
        <td
          className=""
          style={
            Object {
              "fontSize": "1.1em",
              "fontWeight": 700,
              "textAlign": "left",
            }
          }
        >
          SNAP
        </td>
        <td
          className="right aligned"
        >
          $
          2304
           / year
        </td>
        <td
          className="right aligned"
        >
          $
          2304
           / year
        </td>
        <td
          className="right aligned"
        >
          
           $
          0
           / year
        </td>
      </tr>
      <tr
        className=""
      >
        <td
          className=""
          style={
            Object {
              "fontSize": "1.1em",
              "fontWeight": 700,
              "textAlign": "left",
            }
          }
        >
          Section 8 Housing
        </td>
        <td
          className="right aligned"
        >
          $
          0
           / year
        </td>
        <td
          className="right aligned"
        >
          $
          0
           / year
        </td>
        <td
          className="right aligned"
        >
          
           $
          0
           / year
        </td>
      </tr>
      <tr
        className=""
      >
        <td
          className="three wide"
          style={
            Object {
              "borderTop": "2px solid rgba(0, 181, 173, 1)",
              "fontSize": "1.2em",
              "fontWeight": 700,
              "padingTop": "0.25em",
              "textAlign": "left",
            }
          }
        >
          Total Benefits
        </td>
        <td
          className="right aligned three wide"
          style={
            Object {
              "borderTop": "2px solid rgba(0, 181, 173, 1)",
              "fontSize": "1.1em",
              "fontWeight": 700,
              "padingTop": "0.25em",
            }
          }
        >
          $
          2304
           / year
        </td>
        <td
          className="right aligned three wide"
          style={
            Object {
              "borderTop": "2px solid rgba(0, 181, 173, 1)",
              "fontSize": "1.1em",
              "fontWeight": 700,
              "padingTop": "0.25em",
            }
          }
        >
          $
          2304
           / year
        </td>
        <td
          className="right aligned three wide"
          style={
            Object {
              "borderTop": "2px solid rgba(0, 181, 173, 1)",
              "fontSize": "1.1em",
              "fontWeight": 700,
              "padingTop": "0.25em",
            }
          }
        >
          
           $
          0
           / year
        </td>
      </tr>
      <tr
        className=""
      >
        <td
          className=""
          style={
            Object {
              "fontSize": "1.1em",
              "fontWeight": 700,
              "textAlign": "left",
            }
          }
        >
          Income
        </td>
        <td
          className="right aligned"
        >
          $
          0
           / year
        </td>
        <td
          className="right aligned"
        >
          $
          0
           / year
        </td>
        <td
          className="right aligned"
        >
          
           $
          0
           / year
        </td>
      </tr>
      <tr
        className=""
        style={
          Object {
            "border": "none",
          }
        }
      >
        <td
          className="three wide"
          style={
            Object {
              "borderTop": "2px solid rgba(0, 181, 173, 1)",
              "fontSize": "1.2em",
              "fontWeight": 700,
              "padingTop": "0.25em",
              "textAlign": "left",
            }
          }
        >
          Net Total
        </td>
        <td
          className="right aligned three wide"
          style={
            Object {
              "borderTop": "2px solid rgba(0, 181, 173, 1)",
              "fontSize": "1.1em",
              "fontWeight": 700,
              "padingTop": "0.25em",
            }
          }
        >
          $
          2304
           / year
        </td>
        <td
          className="right aligned three wide"
          style={
            Object {
              "borderTop": "2px solid rgba(0, 181, 173, 1)",
              "fontSize": "1.1em",
              "fontWeight": 700,
              "padingTop": "0.25em",
            }
          }
        >
          $
          2304
           / year
        </td>
        <td
          className="right aligned three wide"
          style={
            Object {
              "borderTop": "2px solid rgba(0, 181, 173, 1)",
              "fontSize": "1.1em",
              "fontWeight": 700,
              "padingTop": "0.25em",
            }
          }
        >
          
           $
          0
           / year
        </td>
      </tr>
    </tbody>
  </table>
</wrapper>
`;
